From 1fb52202475a80078cac5c91a87dad4229897c4c Mon Sep 17 00:00:00 2001
From: "SengKai,Tan" <seng.kai.tan@intel.com>
Date: Mon, 4 Jul 2016 18:23:39 +0800
Subject: [PATCH 22/31] x86: aic3107 I2C Interface register for APLI LHCRB

This commit register I2C bus 3 for aic3107 codec specific for
Apollo Lake-I Leaf Hill CRB

Signed-off-by: Tan, Seng Kai <seng.kai.tan@intel.com>
---
 arch/x86/Kconfig                                   |  8 +++
 arch/x86/platform/apli_lhcrb/Makefile              |  1 +
 .../platform/apli_lhcrb/apli_lhcrb_aic3107_i2c.c   | 71 ++++++++++++++++++++++
 3 files changed, 80 insertions(+)
 create mode 100644 arch/x86/platform/apli_lhcrb/apli_lhcrb_aic3107_i2c.c

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 7985fbb..90c8c82 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -680,6 +680,14 @@ config X86_INTEL_APLI_LHCRB_WM8731
 	  bus 3 access for wm8731 codec.
 	  If you do not have one of these chips, you should say N here.
 
+config X86_INTEL_APLI_LHCRB_AIC3107
+	tristate "Intel Apollo Lake-I Leaf Hill I2C Interface for AIC3107 Codec "
+	depends on X86_INTEL_LPSS
+	---help---
+	  This option enables Apollo Lake-I Leaf Hill CRB I2C interface
+	  bus 3 access for aic3107 codec.
+	  If you do not have one of these chips, you should say N here.
+
 # Alphabetically sorted list of Non standard 32 bit platforms
 
 config X86_SUPPORTS_MEMORY_FAILURE
diff --git a/arch/x86/platform/apli_lhcrb/Makefile b/arch/x86/platform/apli_lhcrb/Makefile
index 5812671..f16a2ae 100644
--- a/arch/x86/platform/apli_lhcrb/Makefile
+++ b/arch/x86/platform/apli_lhcrb/Makefile
@@ -1 +1,2 @@
 obj-$(CONFIG_X86_INTEL_APLI_LHCRB_WM8731) += apli_lhcrb_wm8731_i2c.o
+obj-$(CONFIG_X86_INTEL_APLI_LHCRB_AIC3107) += apli_lhcrb_aic3107_i2c.o
diff --git a/arch/x86/platform/apli_lhcrb/apli_lhcrb_aic3107_i2c.c b/arch/x86/platform/apli_lhcrb/apli_lhcrb_aic3107_i2c.c
new file mode 100644
index 0000000..f3174ea
--- /dev/null
+++ b/arch/x86/platform/apli_lhcrb/apli_lhcrb_aic3107_i2c.c
@@ -0,0 +1,71 @@
+/*
+ * I2C Platform initialize for Apollo Lake Machine Driver
+ * Copyright (c) 2016, Intel Corporation.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/i2c.h>
+
+
+#define AIC3107_I2C_ADDRESS	0x18
+#define AUDIO_I2C_CHANNEL	3
+
+static struct i2c_client *client;
+
+static const struct i2c_board_info i2c_bus_info[] __initconst = {
+	{ I2C_BOARD_INFO("tlv320aic3007", AIC3107_I2C_ADDRESS) },
+};
+
+static int __init apli_lhcrb_aic3107_i2c_init(void)
+{
+	int i;
+	struct i2c_adapter *adapter = NULL;
+
+	adapter = i2c_get_adapter(AUDIO_I2C_CHANNEL);
+	if (!adapter) {
+		pr_warn("i2c adapter not found\n");
+		goto exit;
+	}
+
+	client = i2c_new_device(adapter, &i2c_bus_info[0]);
+	if (client == NULL) {
+		pr_warn("i2c new device failed\n");
+		goto exit;
+	}
+
+	i2c_put_adapter(adapter);
+
+	return 0;
+
+exit:
+	i2c_put_adapter(adapter);
+	if (client)
+		i2c_unregister_device(client);
+	return -EIO;
+
+}
+
+static void __exit apli_lhcrb_aic3107_i2c_exit(void)
+{
+	if (client)
+		i2c_unregister_device(client);
+}
+
+device_initcall(apli_lhcrb_aic3107_i2c_init);
+module_exit(apli_lhcrb_aic3107_i2c_exit);
+
+MODULE_DESCRIPTION("Intel Apollo Lake-I Leaf Hill ASoC Machine Driver Init");
+MODULE_LICENSE("GPL v2");
+MODULE_ALIAS("platform:lfcrb_aic3107M_i2s");
+MODULE_ALIAS("platform:lfcrb_aic3107S_i2s");
-- 
1.9.1

